Okta/OktaExternalUserAgentIOS.swift

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2017, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import Foundation
+import OktaAppAuth
+import SafariServices
+
+class OktaExternalUserAgentIOS : NSObject, OIDExternalUserAgent {
+    
+    private let presentingViewController: UIViewController
+    fileprivate var safariViewController: SFSafariViewController?
+    
+    fileprivate var session: OIDExternalUserAgentSession?
+    
+    fileprivate var externalUserAgentFlowInProgress: Bool = false
+    
+    init(presenting presentingViewController: UIViewController) {
+        self.presentingViewController = presentingViewController
+    }
+
+    func present(_ request: OIDExternalUserAgentRequest, session: OIDExternalUserAgentSession) -> Bool {
+        guard !externalUserAgentFlowInProgress else {
+            return false
+        }
+        
+        guard let requestURL = request.externalUserAgentRequestURL() else {
+            return false
+        }
+        self.session = session
+        externalUserAgentFlowInProgress = true
+        var isBrowserLaunched = false
+        
+        if #available(iOS 9.0, *) {
```

Comment:
Also, we new browser APIs were introduced with iOS 11 and 12. We need to use `SFAuthenticationSession` and `ASWebAuthenticationSession`.

Reply:
We are currently using `SFSafariViewController` on iOS 9-12 intentionally due to some problems with how `ASWebAuthenticationSession` is implemented. See https://stackoverflow.com/q/47207914/3191599